utils/zerofrom/src/zero_from.rs

```diff
@@ -0,0 +1,139 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[cfg(feature = "alloc")]
+use alloc::borrow::{Cow, ToOwned};
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+
+/// Trait for types that can be created from a reference to a cart type `C` with no allocations,
+/// i.e. a zero-copy (zero-alloc) version of "From"
+///
+/// A type can be the `ZeroFrom` target of multiple cart types.
+///
+/// The intention is for `ZeroFrom` to produce a struct from a cart with as little work as
+/// possible. Although it is technically possible to implement `ZeroFrom` without being
+/// zero-copy (using heap allocations), doing so defeats the purpose of `ZeroFrom`.
+///
+/// For example, `impl ZeroFrom<C> for Cow<str>` should return a `Cow::Borrowed` pointing at
+/// data in the cart `C`, even if the cart is itself fully owned.
+///
+/// One can use the [`#[derive(ZeroFrom)]`](zerofrom_derive::ZeroFrom) custom derive to automatically
+/// implement this trait.
+///
+/// # Examples
+///
+/// Implementing `ZeroFrom` on a custom data struct:
+///
+/// ```
+/// use zerofrom::ZeroFrom;
+/// use std::borrow::Cow;
+///
+/// struct MyStruct<'data> {
+///     message: Cow<'data, str>,
+/// }
+///
+/// // Reference from a borrowed version of self
+/// impl<'zf> ZeroFrom<'zf, MyStruct<'_>> for MyStruct<'zf> {
+///     fn zero_from(cart: &'zf MyStruct<'_>) -> Self {
+///         MyStruct {
+///             message: Cow::Borrowed(&cart.message)
+///         }
+///     }
+/// }
+///
+/// // Reference from a string slice directly
+/// impl<'zf> ZeroFrom<'zf, str> for MyStruct<'zf> {
+///     fn zero_from(cart: &'zf str) -> Self {
+///         MyStruct {
+///             message: Cow::Borrowed(cart)
+///         }
+///     }
+/// }
+/// ```
+pub trait ZeroFrom<'zf, C: ?Sized>: 'zf {
+    /// Clone the cart `C` into a struct that may retain references into `C`.
+    fn zero_from(cart: &'zf C) -> Self;
```

Comment:
Calling the argument `cart` leaks `Yokeable` concepts into this crate, which is not necessary as `ZeroFrom` can be used independently. The argument should be unnamed like in `From` (and the generic should be `T`).